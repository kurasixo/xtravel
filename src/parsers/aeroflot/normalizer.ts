import type { Price, RawPrice, RawRoute, Route, RouteByName } from '../../types';
import { AdditionalArgsType } from '../parser';


const normalizeTime = (time: string, separator = ':') => {
  return time.split(separator).reduce((acc, el, index) => {
    if (index === 0) {
      acc += +el * 60;
    } else {
      acc += +el;
    }
    return acc;
  }, 0);
};

const normalizeTransferTime = (time: string | null) => {
  if (time === null) {
    return 0;
  }

  const numericalRegex = /[^0-9 ]+/g;
  const timeWithoutLetters = time.replace(numericalRegex, '').trim();
  return normalizeTime(timeWithoutLetters, ' ');
};

const normalizePrice = (rawPriceValue: string | null): number => {
  if (rawPriceValue === null) {
    return 0;
  }

  const numericalRegex = /[^0-9]+/g;
  return +rawPriceValue.replace(numericalRegex, '');
};

export const aeroflotNormalizer = (
  rawRoutes: RawRoute[],
  additionalArgs: AdditionalArgsType,
): RouteByName[] => {
  const routes: Route[] = rawRoutes.map((rawRoute) => {
    const normalizedPrice: Price = {
      value: normalizePrice((rawRoute.price as RawPrice).value),
      currency: (rawRoute.price as RawPrice).currency,
    };

    const normalizedTransfers = rawRoute.transfers.map((rawTransfer) => {
      return {
        ...rawTransfer,
        transferTime: normalizeTransferTime(rawTransfer.transferTime)
      };
    });

    return {
      ...rawRoute,
      price: normalizedPrice,
      transfers: normalizedTransfers,
    };
  });

  const [from, to] = additionalArgs[0].dataForStep;
  const routeName = `${from}_${to}`;

  return [{
    [routeName]: routes,
  }];
};
