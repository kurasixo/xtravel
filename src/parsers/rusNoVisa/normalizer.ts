import compose from 'lodash/fp/compose';
import type { VisaInfo, VisaInfoRaw } from './types';


type WId<O> = Omit<O, 'id'> & { id: VisaInfo['id'] };
type WNDocs<O> = Omit<O, 'docs'> & { docs: VisaInfo['docs'] };
type WNDays<O> = Omit<O, 'daysWithoutVisa'> & { daysWithoutVisa: VisaInfo['daysWithoutVisa'] };

type WIdNWNDocs<O> = WId<WNDocs<O>>;
type WIdNWNDocsNWNDays<O> = WId<WNDocs<WNDays<O>>>;

const addId = (visaInfoRaw: VisaInfoRaw, index: number): WId<VisaInfoRaw> => {
  return {
    id: `${visaInfoRaw.countryName}${index}`,
    ...visaInfoRaw,
  };
};

const normalizeDocs = (
  visaInfoRaw: WId<VisaInfoRaw>,
): WIdNWNDocs<VisaInfoRaw> => {
  const specialSymbol = '—';
  const { docs } = visaInfoRaw;

  const splittedDocs = docs
    .split(specialSymbol)
    .map(el => el.trim())
    .filter(el => el)
    .map(el => `${el[0].toUpperCase()}${el.slice(1)}`);

  return {
    ...visaInfoRaw,
    docs: splittedDocs,
  };
};

const normalizeDays = (
  visaInfoRaw: WIdNWNDocs<VisaInfoRaw>,
): WIdNWNDocsNWNDays<VisaInfoRaw> => {
  const specialSymbol = '/';
  const { daysWithoutVisa } = visaInfoRaw;

  const splittedDays = daysWithoutVisa
    .split(specialSymbol)
    .map(el => el.trim())
    .filter(el => el)
    .map(el => {
      const elAsNumber = +el;
      if (Number.isNaN(elAsNumber)) {
        return el;
      }

      return elAsNumber;
    });

  return {
    ...visaInfoRaw,
    daysWithoutVisa: splittedDays,
  };
};

const NormalizePrice = (visaPrice: string | undefined): string => {
  if (!visaPrice) {
    return 'free';
  }

  let newPrice = visaPrice?.trim();

  if (
    !newPrice ||
    newPrice === '' ||
    newPrice === '-' ||
    newPrice.toLowerCase() === 'бесплатно'
  ) {
    return 'free';
  }

  const newLineRegex = /\n/g;
  newPrice = newPrice.replace(newLineRegex, '');

  const rusRegexWithNum = /[a-я].*[0-9]/g;
  if (rusRegexWithNum.test(newPrice)) { // до 40$ и тд
    const rusRegex = /[a-я]/g;
    newPrice = newPrice.replace(rusRegex, '').trim();
  }

  return newPrice;
};

const getCurrencyByNewPriceString = (newVisaPrice: string): string => {
  const numericalRegex = /[0-9]/;
  const maybeCurrency = newVisaPrice
    .split(numericalRegex)
    .map(el => el.trim())
    .map(el => el.replace(/\//g, ''))
    .filter(el => el);

  const currency = maybeCurrency[maybeCurrency.length - 1];

  return currency;
};

const normalizeRealPrice = (newPrice: string, currency: string): number => {
  const newPriceSplitted = newPrice
    .split(currency);

  const realPrice = newPriceSplitted[0].split('').reduce((acc, char) => {
    if (char === ' ') {
      return acc;
    }

    return acc += char;
  }, '');

  return +realPrice;
};

const normalizePrice = (
  visaInfoRaw: WIdNWNDocsNWNDays<VisaInfoRaw>,
): VisaInfo => {
  const { visaPrice } = visaInfoRaw;

  const newPrice = NormalizePrice(visaPrice);
  if (newPrice === 'free') {
    return {
      ...visaInfoRaw,
      visaPrice: 'free',
    };
  }
  const currency = getCurrencyByNewPriceString(newPrice);
  const realPrice = normalizeRealPrice(newPrice, currency);

  return {
    ...visaInfoRaw,
    visaPrice: {
      currency,
      value: realPrice,
    },
  };
};

const normalize: (
  raw: VisaInfoRaw,
  index: number,
) => VisaInfo = compose(normalizePrice, normalizeDays, normalizeDocs, addId);

export const rusNoVisaNormalizer = (
  rawVisaInfos: VisaInfoRaw[],
): VisaInfo[] => rawVisaInfos.map(normalize);
