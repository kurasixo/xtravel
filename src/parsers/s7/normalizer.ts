import type {
  Route,
  RawRoute,
  RouteByName,

  Transfer,
  RawTransfer,

  Price,
  RawPrice,

  AdditionalArgsType,
} from '../parsers.types';


const normalizePrice = (price: RawPrice | RawPrice[]): Price | Price[] => {
  if (Array.isArray(price)) {
    return price.map((priceItem) => {
      const splittedMagicPrice = priceItem.value.split(' ');
      const currency = splittedMagicPrice[splittedMagicPrice.length - 1];
      const rest = splittedMagicPrice.slice(0, splittedMagicPrice.length - 1);

      return {
        value: +(rest.join('')),
        currency,
      };
    });
  }
  const splittedMagicPrice = price.value.split(' ');
  const currency = splittedMagicPrice[splittedMagicPrice.length - 1];
  const rest = splittedMagicPrice.slice(0, splittedMagicPrice.length - 2);

  return {
    value: +(rest.join('')),
    currency,
  };
};

const normalizeTransfer = (transfer: RawTransfer): Transfer => {
  const ruAlpha = /[а-я]/g;
  const [hours, minutes] = transfer.transferTime
    .replace(ruAlpha, '')
    .split('');

  return {
    ...transfer,
    transferTime: (+hours * 60) + (+minutes),
  };
};

export const s7Normalizer = (
  rawRoutes: RawRoute[],
  additionalArgs: AdditionalArgsType,
): RouteByName[] => {
  const routes: Route[] = rawRoutes.map((rawRoute) => {
    const normalizedPrice = normalizePrice(rawRoute.price);
    const normalizedTransfers = rawRoute.transfers.map(normalizeTransfer);

    return {
      ...rawRoute,
      price: normalizedPrice,
      transfers: normalizedTransfers,
    };
  });

  const [from, to] = additionalArgs[0].dataForStep;
  const routeName = `${from}_${to}`;

  return [{
    [routeName]: routes,
  }];
};
