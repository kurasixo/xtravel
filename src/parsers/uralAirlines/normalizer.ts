import type {
  Flight,

  Route,
  RawRoute,
  RouteByName,

  Transfer,
  RawTransfer,

  Price,
  RawPrice,

  AdditionalArgsType,
} from '../parsers.types';


const normalizeFlight = (flight: Flight): Flight => {
  const parentesisRe = /[()]/g;
  const { fromAirport, toAirport, flightNumber, timeFrom, timeTo } = flight;
  const flightNumberNormal = flightNumber?.split('Рейс ')[1];

  const toAirportName = toAirport.name.replaceAll(parentesisRe, '');
  const fromAirportName = fromAirport.name.replaceAll(parentesisRe, '');

  return {
    ...flight,
    timeTo: timeTo.trim(),
    timeFrom: timeFrom.trim(),
    flightNumber: flightNumberNormal,

    toAirport: {
      ...toAirport,
      name: toAirportName,
    },

    fromAirport: {
      ...fromAirport,
      name: fromAirportName,
    }
  };
};

const normalizeTransfer = (transfer: RawTransfer): Transfer => {
  const ruAlpha = /[а-я]/g;
  const [hours, minutes] = transfer.transferTime
    .replace(ruAlpha, '')
    .split(' ')
    .filter(el => el);

  const newTime = (+hours * 60) + (+minutes);

  return {
    transferTime: newTime,
    transferChange: transfer.transferChange,
  };
};

const normalizePrice = (price: RawPrice | RawPrice[]): Price | Price[] => {
  const ruAlpha = /[а-я]/g;

  if (Array.isArray(price)) {
    return price.map((priceItem) => {
      const splittedMagicPrice = priceItem.value.split(' ')
        .filter((word) => !ruAlpha.test(word));
      const currency = splittedMagicPrice[splittedMagicPrice.length - 1];
      const rest = splittedMagicPrice.slice(0, splittedMagicPrice.length - 1);

      return {
        value: +(rest.join('')),
        currency,
      };
    });
  }

  const splittedMagicPrice = price.value.split(' ')
    .filter((word) => !ruAlpha.test(word));
  const currency = splittedMagicPrice[splittedMagicPrice.length - 1];
  const rest = splittedMagicPrice.slice(0, splittedMagicPrice.length - 2);

  return {
    value: +(rest.join('')),
    currency,
  };
};

const normalizeRoute = (rawRoute: RawRoute): Route => {
  const newFlights = rawRoute.flights.map(normalizeFlight);
  const newTransfers = rawRoute.transfers.map(normalizeTransfer);

  const newPrice = normalizePrice(rawRoute.price);

  return {
    price: newPrice,
    flights: newFlights,
    transfers: newTransfers,
  };
};

export const uralAirlineNormalizer = (
  rawRoutes: RawRoute[],
  additionalArgs: AdditionalArgsType,
): RouteByName[] => {
  const routes = rawRoutes.map(normalizeRoute);
  const [from, to] = additionalArgs[0].dataForStep;
  const routeName = `${from}_${to}`;

  return [{
    [routeName]: routes,
  }];
};
