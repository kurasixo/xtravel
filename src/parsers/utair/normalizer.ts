import type { Flight, Price, RawPrice, RawRoute, RawTransfer, Route, RouteByName, Transfer } from '../../types';
import type { AdditionalArgsType } from '../parser';


const normalizeFlightCompany = (flight: Flight): Flight => {
  return {
    ...flight,
    flightCompany: flight.flightCompany.toLocaleLowerCase(),
  };
};


const normalizeTransfer = (rawTransfer: RawTransfer): Transfer => {
  const ruAlpha = /[а-яА-Я]/g;
  const [hours, , minutes] = rawTransfer.transferTime
    .replace(ruAlpha, '')
    .trim()
    .split(' ');

  return {
    transferTime: (+hours * 60) + (+minutes),
    transferChange: 'unavailable',
  };
};

const normalizePrice = (rawPrice: RawPrice | RawPrice[]): Price | Price[] => {
  if (Array.isArray(rawPrice)) {
    return rawPrice
      .filter(el => el.value !== '-')
      .map((rawPriceItem) => {
        const splittedMagicPrice = rawPriceItem.value.split(String.fromCharCode(160)); // nbsp
        const currency = splittedMagicPrice[splittedMagicPrice.length - 1];
        const rest = splittedMagicPrice.slice(0, splittedMagicPrice.length - 1);

        return {
          value: +(rest.join('')),
          currency,
        };
      });
  }

  const splittedMagicPrice = rawPrice.value.split(' ');
  const currency = splittedMagicPrice[splittedMagicPrice.length - 1];
  const rest = splittedMagicPrice.slice(0, splittedMagicPrice.length - 2);

  return {
    value: +(rest.join('')),
    currency,
  };
};

export const utairNormalizer = (
  rawRoutes: RawRoute[],
  additionalArgs: AdditionalArgsType,
): RouteByName[] => {
  const routes: Route[] = rawRoutes.map(({ flights, transfers, price }) => {
    const normalFlights = flights.map(normalizeFlightCompany);
    const normalTransfers = transfers.map(normalizeTransfer);
    const normalPrice = normalizePrice(price);

    return {
      price: normalPrice,
      transfers: normalTransfers,
      flights: normalFlights,
    };
  });

  const [from, to] = additionalArgs[0].dataForStep;
  const routeName = `${from}_${to}`;

  return [{
    [routeName]: routes,
  }];
};
