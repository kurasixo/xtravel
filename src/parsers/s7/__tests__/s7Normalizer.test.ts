import { getS7Normalizer } from '../normalizer';
import { normalizedFlights26102022, normalizedFlights28102022, flightsMock26102022, flightsMock28102022 } from '../s7tHtml.mock';


describe('getS7Normalizer', () => {
  it('should match snapshot on real data from site "moscow omsk 26 10 2022"', () => {
    const normalizer = getS7Normalizer({ date: 'date', from: 'from', to: 'to' });
    const res = normalizer(flightsMock26102022);

    expect(res).toEqual(normalizedFlights26102022);
  });

  it('should match snapshot on real data from site "moscow omsk 28 10 2022"', () => {
    const normalizer = getS7Normalizer({ date: 'date', from: 'from', to: 'to' });
    const res = normalizer(flightsMock28102022);

    expect(res).toEqual(normalizedFlights28102022);
  });
});

