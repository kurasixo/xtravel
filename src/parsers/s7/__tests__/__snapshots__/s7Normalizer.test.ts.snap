// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getS7Normalizer should match snapshot on real data from site "moscow omsk 26 10 2022" 1`] = `
[
  {
    "from_to": [
      {
        "flights": [],
        "price": [
          {
            "currency": "₽",
            "value": 4900,
          },
          {
            "currency": "₽",
            "value": 7600,
          },
          {
            "currency": "₽",
            "value": 14150,
          },
        ],
        "transfers": [],
      },
      {
        "flights": [],
        "price": [
          {
            "currency": "₽",
            "value": 6680,
          },
          {
            "currency": "₽",
            "value": 9680,
          },
          {
            "currency": "₽",
            "value": 17130,
          },
        ],
        "transfers": [],
      },
      {
        "flights": [],
        "price": [
          {
            "currency": "₽",
            "value": 6968,
          },
          {
            "currency": "₽",
            "value": 11718,
          },
          {
            "currency": "₽",
            "value": 22353,
          },
        ],
        "transfers": [],
      },
      {
        "flights": [],
        "price": [
          {
            "currency": "₽",
            "value": 6968,
          },
          {
            "currency": "₽",
            "value": 11718,
          },
          {
            "currency": "₽",
            "value": 22353,
          },
        ],
        "transfers": [],
      },
    ],
  },
]
`;

exports[`getS7Normalizer should match snapshot on real data from site "moscow omsk 28 10 2022" 1`] = `
[
  {
    "from_to": [
      {
        "flights": [],
        "price": [
          {
            "currency": "₽",
            "value": 6680,
          },
          {
            "currency": "₽",
            "value": 9680,
          },
          {
            "currency": "₽",
            "value": 17130,
          },
        ],
        "transfers": [],
      },
      {
        "flights": [],
        "price": [
          {
            "currency": "₽",
            "value": 14045,
          },
          {
            "currency": "₽",
            "value": 18145,
          },
          {
            "currency": "₽",
            "value": 28895,
          },
        ],
        "transfers": [],
      },
      {
        "flights": [],
        "price": [
          {
            "currency": "₽",
            "value": 14048,
          },
          {
            "currency": "₽",
            "value": 17898,
          },
          {
            "currency": "₽",
            "value": 26848,
          },
        ],
        "transfers": [],
      },
      {
        "flights": [],
        "price": [
          {
            "currency": "₽",
            "value": 11623,
          },
          {
            "currency": "₽",
            "value": 15273,
          },
          {
            "currency": "₽",
            "value": 23623,
          },
        ],
        "transfers": [],
      },
    ],
  },
]
`;
